src/events/controller.py

```diff
@@ -0,0 +1,215 @@
+            quorum_uris = {"controller-quorum-uris": self.charm.state.controller_quorum_uris}
+            self.charm.state.cluster.update(quorum_uris)
+
+            generated_password = self.charm.workload.generate_password()
+
+            if self.charm.state.peer_cluster_orchestrator:
+                self.charm.state.peer_cluster_orchestrator.update(quorum_uris)
+
+                if not self.charm.state.peer_cluster_orchestrator.controller_password:
+                    self.charm.state.peer_cluster_orchestrator.update(
+                        {"controller-password": generated_password}
+                    )
+            elif not self.charm.state.peer_cluster.controller_password:
+                # single mode, controller & leader
+                self.charm.state.cluster.update({"controller-password": generated_password})
+
+            bootstrap_data = {
+                "bootstrap-controller": self.charm.state.bootstrap_controller,
+                "bootstrap-unit-id": str(self.charm.state.kraft_unit_id),
+                "bootstrap-replica-id": self.controller_manager.generate_uuid(),
+            }
+            self.charm.state.cluster.update(bootstrap_data)
+
+    def _format_storages(self) -> None:
+        """Format storages provided relevant keys exist."""
+        if not self.charm.state.kraft_mode:
+            return
+
+        self.broker.config_manager.set_server_properties()
+        if self.charm.state.runs_broker:
+            credentials = self.charm.state.cluster.internal_user_credentials
+        elif self.charm.state.runs_controller:
+            credentials = {
+                self.charm.state.peer_cluster.broker_username: self.charm.state.peer_cluster.broker_password
+            }
+
+        self.controller_manager.format_storages(
+            uuid=self.charm.state.peer_cluster.cluster_uuid,
+            internal_user_credentials=credentials,
+            initial_controllers=f"{self.charm.state.peer_cluster.bootstrap_unit_id}@{self.charm.state.peer_cluster.bootstrap_controller}:{self.charm.state.peer_cluster.bootstrap_replica_id}",
+        )
+
+    def _leader_elected(self, event: LeaderElectedEvent) -> None:
+        if (
+            not self.charm.state.cluster.bootstrap_controller
+            or not self.charm.state.runs_controller
+        ):
+            return
+
+        updated_bootstrap_data = {
+            "bootstrap-controller": self.charm.state.bootstrap_controller,
+            "bootstrap-unit-id": str(self.charm.state.kraft_unit_id),
+            "bootstrap-replica-id": self.charm.state.unit_broker.directory_id,
+        }
+        self.charm.state.cluster.update(updated_bootstrap_data)
+
+        if self.charm.state.peer_cluster_orchestrator:
+            self.charm.state.peer_cluster_orchestrator.update(updated_bootstrap_data)
+
+        # change bootstrap controller config on followers and brokers
+        self.charm.on.config_changed.emit()
+
+    def _add_controller(self) -> None:
```

Comment:
**nit**: naming and scope on these two methods should be consistent:
- Either both are private or both are public (leading underscore on the name)
- Either `add_controller | remove_controller` or `add_to_quorum | remove_from_quorum`